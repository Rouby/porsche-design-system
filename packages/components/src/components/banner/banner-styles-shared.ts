import type { JssStyle } from 'jss';
import { addImportantToRule } from '../../styles';

const easeInQuad = 'cubic-bezier(0.45,0,0.55,1)';
const easeOutQuad = 'cubic-bezier(0.5,1,0.89,1)';
export const ANIMATION_DURATION = 600;

export const getAnimationIn = (keyframesName: string, durationVar?: string): JssStyle => {
  const duration = durationVar ? `var(${durationVar},${ANIMATION_DURATION}ms)` : `${ANIMATION_DURATION}ms`;
  return {
    animation: `${duration} $${keyframesName} ${easeInQuad} forwards`,
  };
};

export const getAnimationOut = (keyframesName: string): JssStyle => ({
  animation: addImportantToRule(`${ANIMATION_DURATION}ms $${keyframesName} ${easeOutQuad} forwards`),
});

export type KeyframesDirection = 'in' | 'out';
export const getKeyframes = (direction: KeyframesDirection, outsideStyle: JssStyle): JssStyle => {
  const insideStyle: JssStyle = { opacity: 1, transform: 'translate3d(0,0,0)' };
  return direction === 'in'
    ? {
        from: outsideStyle,
        to: insideStyle,
      }
    : {
        from: insideStyle,
        to: outsideStyle,
      };
};

export const getKeyframesMobile = (direction: KeyframesDirection, bottomVar: string): JssStyle =>
  getKeyframes(direction, {
    opacity: 0,
    transform: `translate3d(0,calc(var(${bottomVar}) + 100%),0)`, // space before and after "+" is crucial
  });
