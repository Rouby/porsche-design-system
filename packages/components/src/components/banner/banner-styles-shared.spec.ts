import { getAnimationIn, getAnimationOut, getKeyframesMobile, KeyframesDirection } from './banner-styles-shared';

describe('getAnimationIn()', () => {
  it.each<string[]>([
    ['mobileIn', '--p-duration'],
    ['mobileIn', null],
  ])('should return correct JssStyle for keyframesName: %s and durationVar: %s', (keyframesName, durationVar) => {
    expect(getAnimationIn(keyframesName, durationVar)).toMatchSnapshot();
  });
});

describe('getAnimationOut()', () => {
  it('should return correct JssStyle', () => {
    expect(getAnimationOut('mobileOut')).toMatchSnapshot();
  });
});

describe('getKeyframesMobile()', () => {
  it.each<[KeyframesDirection, string]>([
    ['in', '--p-bottom-var'],
    ['out', '--p-bottom-var'],
  ])('should return correct JssStyle for direction: %s and bottomVar: %s', (direction, bottomVar) => {
    expect(getKeyframesMobile(direction, bottomVar)).toMatchSnapshot();
  });
});
