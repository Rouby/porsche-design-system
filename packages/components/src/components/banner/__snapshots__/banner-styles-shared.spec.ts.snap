// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAnimationIn() should return correct JssStyle for keyframesName: mobileIn and durationVar: --p-duration 1`] = `
{
  "animation": "var(--p-duration,600ms) $mobileIn cubic-bezier(0.45,0,0.55,1) forwards",
}
`;

exports[`getAnimationIn() should return correct JssStyle for keyframesName: mobileIn and durationVar: null 1`] = `
{
  "animation": "600ms $mobileIn cubic-bezier(0.45,0,0.55,1) forwards",
}
`;

exports[`getAnimationOut() should return correct JssStyle 1`] = `
{
  "animation": "600ms $mobileOut cubic-bezier(0.5,1,0.89,1) forwards !important",
}
`;

exports[`getKeyframesMobile() should return correct JssStyle for direction: in and bottomVar: --p-bottom-var 1`] = `
{
  "from": {
    "opacity": 0,
    "transform": "translate3d(0,calc(var(--p-bottom-var) + 100%),0)",
  },
  "to": {
    "opacity": 1,
    "transform": "translate3d(0,0,0)",
  },
}
`;

exports[`getKeyframesMobile() should return correct JssStyle for direction: out and bottomVar: --p-bottom-var 1`] = `
{
  "from": {
    "opacity": 1,
    "transform": "translate3d(0,0,0)",
  },
  "to": {
    "opacity": 0,
    "transform": "translate3d(0,calc(var(--p-bottom-var) + 100%),0)",
  },
}
`;
